src/pipeline_migration/utils.py

```diff
@@ -1,17 +1,100 @@
+from dataclasses import dataclass, field
 from typing import Any
-from ruamel.yaml import YAML
 
 from pipeline_migration.types import FilePath
 
+from ruamel.yaml import YAML, CommentedMap, CommentedSeq
+from ruamel.yaml.comments import CommentedBase
+
+
+__all__ = [
+    "BlockSequenceIndentation",
+    "create_yaml_obj",
+    "dump_yaml",
+    "is_true",
+    "load_yaml",
+    "YAMLStyle",
+]
+
 
 def is_true(value: str) -> bool:
     return value.strip().lower() == "true"
 
 
-def create_yaml_obj():
+@dataclass
+class BlockSequenceIndentation:
+    indentations: dict[int, int] = field(default_factory=dict)
+
+    @property
+    def is_consistent(self) -> bool:
+        return len(self.levels) == 1
+
+    @property
+    def levels(self) -> list[int]:
+        return list(self.indentations.keys())
+
+
+@dataclass
+class _NodePath:
+    node: CommentedBase
+    field: str = ""
+
+
+@dataclass
+class YAMLStyle:
+    indentation: BlockSequenceIndentation
+
+    preserve_quotes: bool = True
+    width: int = 8192
+
+    @classmethod
+    def _detect_block_sequence_indentation(cls, node: CommentedBase) -> BlockSequenceIndentation:
+
+        parent_nodes: list[_NodePath] = []
+        block_seq_indentations = BlockSequenceIndentation()
+        indentations = block_seq_indentations.indentations
+
+        def _walk(node: CommentedBase) -> None:
+            if isinstance(node, CommentedMap):
+                parent_nodes.append(_NodePath(node=node))
+                for key, value in node.items():
+                    parent_nodes[-1].field = key
+                    _walk(value)
+                parent_nodes.pop()
+            elif isinstance(node, CommentedSeq):
+                levels = node.lc.col - parent_nodes[-1].node.lc.col
+                if levels in indentations:
+                    indentations[levels] += 1
+                else:
+                    indentations[levels] = 1
+                for item in node:
+                    _walk(item)
+
+        _walk(node)
+
+        return block_seq_indentations
+
+    @classmethod
+    def detect(cls, file_path: FilePath) -> "YAMLStyle":
+        doc = load_yaml(file_path)
+        indentation = cls._detect_block_sequence_indentation(doc)
+        return cls(indentation=indentation)
+
+
+def create_yaml_obj(style: YAMLStyle | None = None):
     yaml = YAML()
-    yaml.preserve_quotes = True
-    yaml.width = 8192
+    if style is None:
+        return yaml
+
+    yaml.preserve_quotes = style.preserve_quotes
+    yaml.width = style.width
+
+    offset = 0
+    if style.indentation.is_consistent:
+        offset = style.indentation.levels[0]
```

Comment:
I notice we're already counting the occurrences for each indentation level. How about, instead of checking consistency, we just use the most common indentation level?

That way, if most of the file is indented with e.g. 2 spaces and someone adds a small incorrectly indented part, we'll only reformat the small part instead of reformatting the whole file.

Implementation detail: it could be nice to use [`collections.Counter`](https://docs.python.org/3/library/collections.html#collections.Counter) for this to 1) better indicate the intent and 2) make use of the `most_common()` method

Reply:
I think the idea behind my comment to STONEBLD-3422 is also applied here. The key point is we have to provide an explicit Konflux behavior than implicit. There is another story for build-service to prompt users in onboard PR for using consistent indentation style. Then, Konflux proposes consistent indentation formatting and users are responsible for maintaining for it.

The number of occurrences for each indentation level as well as the `_NodePath.field` are mostly for debug purpose. The code is extensible for verbose output.

Reply:
Ok, sounds reasonable